### PR TITLE
Fix dgc buffer illegal & reuse velocity

### PR DESCRIPTION
**Pull request by wangxicoding**
1. Fix DGC gather buffer illegal problem by alloc buffer in dgc_op, avoid multi-stream bug in memory::Alloc
2. Reuse velocity in dgc_op & dgc_momentum_op, save memory & reserve global momentum information.

### CODE REVIEW

File: paddle/fluid/framework/details/sparse_all_reduce_op_handle.cc [hunk 1]
```diff
@@ -135,9 +137,11 @@ void SparseAllReduceOpHandle::RunImplEncoded() {
     // dgc use ncclAllGather to get all the encoded data
     // so the buffer need nranks.
     int buf_size = nranks_ * encode_size;
-    auto tmp_ious_data = memory::Alloc(place, buf_size);
-    void *gather_buff = reinterpret_cast<void *>(tmp_ious_data->ptr());
-    allocations.emplace_back(std::move(tmp_ious_data));
+    // auto tmp_ious_data = memory::Alloc(place, buf_size);
```
Comment (gongweibao) [inline paddle/fluid/framework/details/sparse_all_reduce_op_handle.cc]: clean up these codes!

**Review by gongweibao — CHANGES_REQUESTED** [review]
Small needs fix!

File: paddle/fluid/framework/details/sparse_all_reduce_op_handle.cc (hunk 1 above)
Comment (wangxicoding) [reply paddle/fluid/framework/details/sparse_all_reduce_op_handle.cc]: Done.

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**